File: cartalk/protocol/uds.py
"""UDS (ISO 14229) client.

A thin client over a Transport. The services the scanner needs are implemented directly
against the byte protocol so the core works without any external UDS library; richer
features (Phase 3 writes, security access) can delegate to ``udsoncan`` later.
"""

from __future__ import annotations

import time

from .dtc import Dtc, decode_dtc_block

# UDS service request SIDs; positive response SID = request SID + 0x40.
SID_DIAGNOSTIC_SESSION_CONTROL = 0x10
SID_CLEAR_DIAGNOSTIC_INFORMATION = 0x14
SID_READ_DTC_INFORMATION = 0x19
SID_READ_DATA_BY_IDENTIFIER = 0x22
SID_WRITE_DATA_BY_IDENTIFIER = 0x2E
SID_SECURITY_ACCESS = 0x27
SID_TESTER_PRESENT = 0x3E

NEGATIVE_RESPONSE = 0x7F
NRC_RESPONSE_PENDING = 0x78

# ReadDTCInformation sub-functions.
RDTC_BY_STATUS_MASK = 0x02          # reportDTCByStatusMask
STATUS_MASK_ALL = 0xFF

# DiagnosticSessionControl sub-functions.
SESSION_DEFAULT = 0x01
SESSION_EXTENDED = 0x03


class UdsError(Exception):
    """A UDS negative response or protocol violation."""

    def __init__(self, message: str, nrc: int | None = None):
        super().__init__(message)
        self.nrc = nrc


class UdsClient:
    """UDS requests against one module (request/response CAN id pair)."""

    def __init__(self, transport, request_id: int, response_id: int,
                 timeout: float = 2.0, pending_timeout: float = 5.0):
        self.transport = transport
        self.request_id = request_id
        self.response_id = response_id
        self.timeout = timeout
        # Max wall-clock to keep re-requesting while the server answers 0x78.
        self.pending_timeout = pending_timeout

    def _send(self, payload: bytes) -> bytes:
        """Send a request, transparently retrying while the server replies 0x78.

        A 0x78 ("requestCorrectlyReceived-ResponsePending") means the ECU needs more
        time. The transport couples send+receive, so we re-issue the (idempotent) request
        until we get a non-pending answer or exhaust ``pending_timeout``.
        """
        deadline = time.monotonic() + self.pending_timeout
        while True:
            resp = self.transport.request(self.request_id, self.response_id, payload,
                                          timeout=self.timeout)
            if not resp:
                raise UdsError("empty response")
            if resp[0] == NEGATIVE_RESPONSE:
                nrc = resp[2] if len(resp) >= 3 else None
                if nrc == NRC_RESPONSE_PENDING and time.monotonic() < deadline:
                    time.sleep(0.05)
                    continue
                raise UdsError(
                    f"negative response to {resp[1]:#04x}, NRC {nrc:#04x}"
                    if nrc is not None else "negative response", nrc=nrc)
            return resp

    def _expect(self, payload: bytes, sid: int) -> bytes:
        resp = self._send(payload)
        if resp[0] != sid + 0x40:
            raise UdsError(f"unexpected response SID {resp[0]:#04x}, wanted {sid + 0x40:#04x}")
        return resp

    # -- services -----------------------------------------------------------

    def diagnostic_session(self, session: int = SESSION_EXTENDED) -> bytes:
        """DiagnosticSessionControl (0x10)."""
        return self._expect(bytes([SID_DIAGNOSTIC_SESSION_CONTROL, session]),
                            SID_DIAGNOSTIC_SESSION_CONTROL)

    def tester_present(self) -> None:
        """TesterPresent (0x3E) with suppressPositiveResponse — keepalive."""
        # 0x80 sets the suppress bit, so the ECU stays in session without replying.
        self.transport.request(self.request_id, self.response_id,
                               bytes([SID_TESTER_PRESENT, 0x80]), timeout=self.timeout)

    def read_dtcs(self, status_mask: int = STATUS_MASK_ALL) -> list[Dtc]:
        """ReadDTCInformation / reportDTCByStatusMask (0x19 0x02)."""
        req = bytes([SID_READ_DTC_INFORMATION, RDTC_BY_STATUS_MASK, status_mask])
        resp = self._expect(req, SID_READ_DTC_INFORMATION)
        # resp: [0x59, 0x02, statusAvailabilityMask, <4-byte DTC records...>]
        body = resp[3:]
        return decode_dtc_block(body, record_size=4)

    def read_did(self, did: int) -> bytes:
        """ReadDataByIdentifier (0x22). Returns the raw data bytes for the DID."""
        req = bytes([SID_READ_DATA_BY_IDENTIFIER, (did >> 8) & 0xFF, did & 0xFF])
        resp = self._expect(req, SID_READ_DATA_BY_IDENTIFIER)
        # resp: [0x62, did_hi, did_lo, <data...>]
        return resp[3:]

    def clear_dtcs(self, group: int = 0xFFFFFF) -> None:
        """ClearDiagnosticInformation (0x14). Requires SGW bypass on 2018+ FCA.

        Not wired into any user-facing command yet — clearing is a write and waits for
        the Phase 3 write-safety gate.
        """
        req = bytes([SID_CLEAR_DIAGNOSTIC_INFORMATION,
                     (group >> 16) & 0xFF, (group >> 8) & 0xFF, group & 0xFF])
        self._expect(req, SID_CLEAR_DIAGNOSTIC_INFORMATION)

File: web/lib/uds.js
// UDS (ISO 14229) client — port of cartalk/protocol/uds.py.
// Operates over a `send(reqId, respId, payloadUint8) -> Promise<Uint8Array>` function so it
// is transport-agnostic (real ELM327/WebUSB in the browser; a fake in tests).

import { decodeDtcBlock } from "./dtc.js";

export const SID_DIAGNOSTIC_SESSION_CONTROL = 0x10;
export const SID_READ_DTC_INFORMATION = 0x19;
export const SID_READ_DATA_BY_IDENTIFIER = 0x22;
export const SID_TESTER_PRESENT = 0x3e;
export const NEGATIVE_RESPONSE = 0x7f;
export const NRC_RESPONSE_PENDING = 0x78;
export const RDTC_BY_STATUS_MASK = 0x02;
export const STATUS_MASK_ALL = 0xff;
export const SESSION_EXTENDED = 0x03;

export class UdsError extends Error {
  constructor(message, nrc = null) {
    super(message);
    this.name = "UdsError";
    this.nrc = nrc;
  }
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

export class UdsClient {
  /**
   * @param {(reqId:number,respId:number,payload:Uint8Array)=>Promise<Uint8Array>} send
   */
  constructor(send, reqId, respId, { pendingTimeoutMs = 5000 } = {}) {
    this.send = send;
    this.reqId = reqId;
    this.respId = respId;
    this.pendingTimeoutMs = pendingTimeoutMs;
  }

  async _send(payload) {
    const deadline = Date.now() + this.pendingTimeoutMs;
    for (;;) {
      const resp = await this.send(this.reqId, this.respId, payload);
      if (!resp || resp.length === 0) throw new UdsError("empty response");
      if (resp[0] === NEGATIVE_RESPONSE) {
        const nrc = resp.length >= 3 ? resp[2] : null;
        if (nrc === NRC_RESPONSE_PENDING && Date.now() < deadline) {
          await sleep(50);
          continue;
        }
        throw new UdsError(
          nrc !== null
            ? `negative response to 0x${(resp[1] || 0).toString(16)}, NRC 0x${nrc.toString(16)}`
            : "negative response",
          nrc
        );
      }
      return resp;
    }
  }

  async _expect(payload, sid) {
    const resp = await this._send(payload);
    if (resp[0] !== sid + 0x40) {
      throw new UdsError(`unexpected response SID 0x${resp[0].toString(16)}, wanted 0x${(sid + 0x40).toString(16)}`);
    }
    return resp;
  }

  async diagnosticSession(session = SESSION_EXTENDED) {
    return this._expect(Uint8Array.from([SID_DIAGNOSTIC_SESSION_CONTROL, session]),
                        SID_DIAGNOSTIC_SESSION_CONTROL);
  }

  async testerPresent() {
    await this.send(this.reqId, this.respId, Uint8Array.from([SID_TESTER_PRESENT, 0x80]));
  }

  async readDtcs(statusMask = STATUS_MASK_ALL) {
    const req = Uint8Array.from([SID_READ_DTC_INFORMATION, RDTC_BY_STATUS_MASK, statusMask]);
    const resp = await this._expect(req, SID_READ_DTC_INFORMATION);
    return decodeDtcBlock(resp.slice(3), 4);  // skip [0x59, 0x02, availabilityMask]
  }

  async readDid(did) {
    const req = Uint8Array.from([SID_READ_DATA_BY_IDENTIFIER, (did >> 8) & 0xff, did & 0xff]);
    const resp = await this._expect(req, SID_READ_DATA_BY_IDENTIFIER);
    return resp.slice(3);  // skip [0x62, didHi, didLo]
  }
}
